calico/openstack/mech_calico.py

```diff
@@ -600,13 +614,20 @@ def _update_port(self, context, port):
         if not port_disabled:
             LOG.info("Port enabled, attempting to update.")
 
-            with context._plugin_context.session.begin(subtransactions=True):
-                port = self.db.get_port(context._plugin_context, port['id'])
+            # When called directly from the core plugin, we don't have the ML2
+            # plugin context.
+            if getattr(context, '_plugin_context', None):
```

Comment:
`_plugin_context` is a kind of context that is used throughout the OpenStack codebase, that describes both database context, e.g. any open transactions, and request context i.e. what the overall REST API request is that is driving the code currently being executed.

`context` is an ML2 superset of that.  In addition to the above, it contains fields like the old and new port dict, if doing an update_port_* call.

Does that help?